src/libraries/System.Private.CoreLib/src/System/Text/ASCIIUtility.cs

```diff
@@ -215,17 +215,54 @@ private static unsafe nuint GetIndexOfFirstNonAsciiByte_Default(byte* pBuffer, n
             goto Finish;
         }
 
-        private static unsafe nuint GetIndexOfFirstNonAsciiByte_Sse2(byte* pBuffer, nuint bufferLength)
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static ulong GetMaskOfMostSignificantBits(Vector128<byte> value)
+        {
+            Debug.Assert(BitConverter.IsLittleEndian);
+            if (!AdvSimd.Arm64.IsSupported)
+            {
+                throw new PlatformNotSupportedException();
+            }
+
+            // extractedBits[i] = (value[i] & 0x80 == 0x80) & (1 << (12 * (i % 2)))
+            Vector128<byte> s_mostSignficantBitMask = Vector128.Create((byte)0x80);
+            Vector128<byte> s_bitmask = Vector128.Create((ushort)0x1001).AsByte();
+            Vector128<byte> mostSignificantBitIsSet = AdvSimd.CompareTest(value, s_mostSignficantBitMask);
+            Vector128<byte> extractedBits = AdvSimd.And(mostSignificantBitIsSet, s_bitmask);
+
+            extractedBits = AdvSimd.Arm64.AddPairwise(extractedBits, extractedBits);
+            extractedBits = AdvSimd.Arm64.AddPairwise(extractedBits, extractedBits);
+            extractedBits = AdvSimd.Arm64.AddPairwise(extractedBits, extractedBits);
+            return extractedBits.AsUInt64().ToScalar();
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static bool ContainsNonAsciiByte(Vector128<ushort> value)
```

Comment:
>ContainsNonAsciiByte [](start = 27, length = 20)

can you confirm what needs to happen here? `ushort` is 16 bits long and you want to check if bit-15 of any lane is set on not. In that case, the mask `0x8080808080808080` which is meant for `byte` won't work here. You might need `0x8000800080008000`.

Reply:
Yup, you're right. I missed that. Fixed.